File: backend/src/dcim/services/power_chain.py
"""Power-chain rollup for a rack.

Returns:
  per_asset: { asset_id: {
      sides_covered: ["A","B"],          # which PDU sides this device is fed from
      connections: [
        { pdu_id, pdu_name, pdu_side, outlet_id, outlet_position, psu_index }
      ],
      redundancy: "redundant" | "single" | "unpowered" | "n/a"
    }
  }
  pdus: [ { id, name, side, total_outlets, used_outlets } ]

Redundancy rules (rules of thumb that match real DC operations):
  - PDU asset itself: n/a
  - Device with no power_count constraint and no connections: unpowered
  - Device fed from outlets on 2+ different PDU sides: redundant
  - Device fed from at least one outlet but all on the same side: single
"""

from __future__ import annotations

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ..models.inventory import Asset, AssetKind
from ..models.power import Outlet, PowerConnection


def classify_redundancy(kind, connections: list[dict]) -> tuple[list[str], str]:
    """Pure classifier — given a device kind + its power connections, decide
    sides_covered and the redundancy verdict. Extracted so tests don't need a DB.

    `kind` is compared by value-equality to AssetKind.pdu (str enum) so the
    function works whether callers pass the enum, the string "pdu", or a
    SimpleNamespace stand-in.
    """
    pdu_value = AssetKind.pdu.value if hasattr(AssetKind.pdu, "value") else AssetKind.pdu
    if kind == AssetKind.pdu or kind == pdu_value:
        return [], "n/a"
    sides = sorted({c["pdu_side"] for c in connections if c.get("pdu_side")})
    if not connections:
        return sides, "unpowered"
    if len(sides) >= 2:
        return sides, "redundant"
    return sides, "single"


async def _load_outlets_and_connections(
    db: AsyncSession, pdu_ids: list,
) -> tuple[list[Outlet], list[PowerConnection]]:
    if not pdu_ids:
        return [], []
    outlets = list(
        (
            await db.execute(select(Outlet).where(Outlet.pdu_asset_id.in_(pdu_ids)))
        ).scalars().all()
    )
    if not outlets:
        return [], []
    conns = list(
        (
            await db.execute(
                select(PowerConnection).where(
                    PowerConnection.outlet_id.in_([o.id for o in outlets])
                )
            )
        ).scalars().all()
    )
    return outlets, conns


def _connection_row(pdu: Asset, outlet: Outlet, conn: PowerConnection) -> dict:
    return {
        "pdu_id": str(pdu.id),
        "pdu_name": pdu.name,
        "pdu_side": pdu.pdu_side.value if pdu.pdu_side else None,
        "outlet_id": str(outlet.id),
        "outlet_position": outlet.position,
        "outlet_label": outlet.label,
        "psu_index": conn.psu_index,
    }


def _empty_entry() -> dict:
    return {"sides_covered": [], "connections": [], "redundancy": "n/a"}


def _pdu_summary_row(p: Asset, outlets_for_p: list[Outlet], used_outlet_ids: set) -> dict:
    return {
        "id": str(p.id),
        "name": p.name,
        "side": p.pdu_side.value if p.pdu_side else None,
        "mount": p.mount.value if hasattr(p.mount, "value") else p.mount,
        "face": p.face.value if hasattr(p.face, "value") else p.face,
        "total_outlets": len(outlets_for_p),
        "used_outlets": sum(1 for o in outlets_for_p if o.id in used_outlet_ids),
    }


async def compute_power_chain(db: AsyncSession, assets: list[Asset]) -> dict:
    if not assets:
        return {"per_asset": {}, "pdus": []}

    pdus = [a for a in assets if a.kind == AssetKind.pdu]
    outlets_rows, conns = await _load_outlets_and_connections(db, [p.id for p in pdus])
    outlets_by_id = {o.id: o for o in outlets_rows}
    outlets_by_pdu: dict[str, list[Outlet]] = {}
    for o in outlets_rows:
        outlets_by_pdu.setdefault(str(o.pdu_asset_id), []).append(o)
    pdu_by_id = {str(p.id): p for p in pdus}

    per_asset: dict[str, dict] = {str(a.id): _empty_entry() for a in assets}
    for c in conns:
        outlet = outlets_by_id.get(c.outlet_id)
        pdu = pdu_by_id.get(str(outlet.pdu_asset_id)) if outlet else None
        if outlet is None or pdu is None:
            continue
        entry = per_asset.setdefault(str(c.asset_id), _empty_entry())
        entry["connections"].append(_connection_row(pdu, outlet, c))

    for a in assets:
        entry = per_asset[str(a.id)]
        sides, verdict = classify_redundancy(a.kind, entry["connections"])
        entry["sides_covered"] = sides
        entry["redundancy"] = verdict

    used_outlet_ids = {c.outlet_id for c in conns}
    pdu_summary = [
        _pdu_summary_row(p, outlets_by_pdu.get(str(p.id), []), used_outlet_ids)
        for p in pdus
    ]
    return {"per_asset": per_asset, "pdus": pdu_summary}

File: packages/otter-go/internal/powerchain/powerchain.go
// Package powerchain holds the rack-level power-graph rollup used by
// the rack-detail dashboard. Mirrors packages/otter/src/dcim/services/
// power_chain.py byte-for-byte. Walks the asset → PDU → outlet graph
// to produce the per-asset connection list + redundancy verdict and
// the PDU summary's used-outlets counts.
package powerchain

import (
	"context"

	"github.com/google/uuid"

	dbq "github.com/usg-dcim/packages/otter-go/db/generated"
)

// Result is the {per_asset, pdus} wire shape from compute_power_chain.
// Maps stay keyed on string-stringified UUID to match the Python dict
// keys (finch JSON parses them as strings).
type Result struct {
	PerAsset map[string]AssetEntry `json:"per_asset"`
	PDUs     []PduSummary          `json:"pdus"`
}

// AssetEntry is the per-asset rollup — which PDU sides feed it,
// the populated connections, and the redundancy verdict.
type AssetEntry struct {
	SidesCovered []string     `json:"sides_covered"`
	Connections  []Connection `json:"connections"`
	Redundancy   string       `json:"redundancy"`
}

// Connection mirrors Python's _connection_row — id+name of the PDU,
// the side label, the outlet identity, and the psu_index of the
// attached asset.
type Connection struct {
	PduID          string  `json:"pdu_id"`
	PduName        string  `json:"pdu_name"`
	PduSide        *string `json:"pdu_side"`
	OutletID       string  `json:"outlet_id"`
	OutletPosition int32   `json:"outlet_position"`
	OutletLabel    *string `json:"outlet_label"`
	PsuIndex       int32   `json:"psu_index"`
}

// PduSummary mirrors Python's _pdu_summary_row.
type PduSummary struct {
	ID           string  `json:"id"`
	Name         string  `json:"name"`
	Side         *string `json:"side"`
	Mount        string  `json:"mount"`
	Face         string  `json:"face"`
	TotalOutlets int     `json:"total_outlets"`
	UsedOutlets  int     `json:"used_outlets"`
}

// Querier is the slice of dbq methods Compute needs. The existing
// ListPowerConnectionsByOutletIDs (returns []PowerConnection) is
// reused; ListOutletsByPduIDs is the new bulk lookup.
type Querier interface {
	ListOutletsByPduIDs(ctx context.Context, pduIDs []uuid.UUID) ([]dbq.OutletForPowerChainRow, error)
	ListPowerConnectionsByOutletIDs(ctx context.Context, outletIDs []uuid.UUID) ([]dbq.PowerConnection, error)
}

// Compute returns the power-chain rollup for the given assets. Mirrors
// services/power_chain.py::compute_power_chain.
func Compute(ctx context.Context, q Querier, assets []dbq.Asset) (Result, error) {
	if len(assets) == 0 {
		return Result{PerAsset: map[string]AssetEntry{}, PDUs: []PduSummary{}}, nil
	}
	pdus := pduAssetsOf(assets)
	outlets, conns, err := loadOutletsAndConnections(ctx, q, pdus)
	if err != nil {
		return Result{}, err
	}
	outletByID, outletsByPdu := indexOutlets(outlets)
	pduByID := indexPdusByID(pdus)
	perAsset := emptyPerAsset(assets)

	for _, c := range conns {
		outlet, okO := outletByID[c.OutletID]
		if !okO {
			continue
		}
		pdu, okP := pduByID[outlet.PduAssetID]
		if !okP {
			continue
		}
		key := c.AssetID.String()
		entry := perAsset[key]
		if entry.Connections == nil {
			entry = emptyEntry()
		}
		entry.Connections = append(entry.Connections, buildConnection(pdu, outlet, c))
		perAsset[key] = entry
	}

	for _, a := range assets {
		key := a.ID.String()
		entry := perAsset[key]
		sides, verdict := ClassifyRedundancy(a.Kind, entry.Connections)
		entry.SidesCovered = sides
		entry.Redundancy = verdict
		perAsset[key] = entry
	}

	usedOutletIDs := usedOutletIDsFrom(conns)
	pduSummary := make([]PduSummary, 0, len(pdus))
	for _, p := range pdus {
		pduSummary = append(pduSummary, buildPduSummary(p, outletsByPdu[p.ID], usedOutletIDs))
	}
	return Result{PerAsset: perAsset, PDUs: pduSummary}, nil
}

// ClassifyRedundancy is the pure classifier — given a device's kind
// and its connections, return (sides_covered, verdict). Extracted so
// tests don't need a DB. Mirrors services/power_chain.py exactly:
//
//	PDU itself                            → ([], "n/a")
//	no connections                        → (sides_seen, "unpowered")
//	≥2 distinct PDU sides                 → (sides_seen, "redundant")
//	≥1 connection, single PDU side seen   → (sides_seen, "single")
func ClassifyRedundancy(kind string, connections []Connection) ([]string, string) {
	if kind == "pdu" {
		return []string{}, "n/a"
	}
	sides := dedupedSides(connections)
	if len(connections) == 0 {
		return sides, "unpowered"
	}
	if len(sides) >= 2 {
		return sides, "redundant"
	}
	return sides, "single"
}

// ---- internal helpers ----

func pduAssetsOf(assets []dbq.Asset) []dbq.Asset {
	out := []dbq.Asset{}
	for _, a := range assets {
		if a.Kind == "pdu" {
			out = append(out, a)
		}
	}
	return out
}

func loadOutletsAndConnections(ctx context.Context, q Querier, pdus []dbq.Asset) (
	[]dbq.OutletForPowerChainRow, []dbq.PowerConnection, error,
) {
	if len(pdus) == 0 {
		return nil, nil, nil
	}
	pduIDs := make([]uuid.UUID, len(pdus))
	for i, p := range pdus {
		pduIDs[i] = p.ID
	}
	outlets, err := q.ListOutletsByPduIDs(ctx, pduIDs)
	if err != nil {
		return nil, nil, err
	}
	if len(outlets) == 0 {
		return outlets, nil, nil
	}
	outletIDs := make([]uuid.UUID, len(outlets))
	for i, o := range outlets {
		outletIDs[i] = o.ID
	}
	conns, err := q.ListPowerConnectionsByOutletIDs(ctx, outletIDs)
	if err != nil {
		return nil, nil, err
	}
	return outlets, conns, nil
}

func indexOutlets(outlets []dbq.OutletForPowerChainRow) (
	map[uuid.UUID]dbq.OutletForPowerChainRow, map[uuid.UUID][]dbq.OutletForPowerChainRow,
) {
	byID := make(map[uuid.UUID]dbq.OutletForPowerChainRow, len(outlets))
	byPdu := map[uuid.UUID][]dbq.OutletForPowerChainRow{}
	for _, o := range outlets {
		byID[o.ID] = o
		byPdu[o.PduAssetID] = append(byPdu[o.PduAssetID], o)
	}
	return byID, byPdu
}

func indexPdusByID(pdus []dbq.Asset) map[uuid.UUID]dbq.Asset {
	out := make(map[uuid.UUID]dbq.Asset, len(pdus))
	for _, p := range pdus {
		out[p.ID] = p
	}
	return out
}

func emptyPerAsset(assets []dbq.Asset) map[string]AssetEntry {
	out := make(map[string]AssetEntry, len(assets))
	for _, a := range assets {
		out[a.ID.String()] = emptyEntry()
	}
	return out
}

func emptyEntry() AssetEntry {
	return AssetEntry{
		SidesCovered: []string{},
		Connections:  []Connection{},
		Redundancy:   "n/a",
	}
}

func buildConnection(pdu dbq.Asset, outlet dbq.OutletForPowerChainRow, c dbq.PowerConnection) Connection {
	return Connection{
		PduID:          pdu.ID.String(),
		PduName:        pdu.Name,
		PduSide:        pdu.PduSide,
		OutletID:       outlet.ID.String(),
		OutletPosition: outlet.Position,
		OutletLabel:    outlet.Label,
		PsuIndex:       c.PsuIndex,
	}
}

func buildPduSummary(p dbq.Asset, outletsForP []dbq.OutletForPowerChainRow, usedOutletIDs map[uuid.UUID]struct{}) PduSummary {
	used := 0
	for _, o := range outletsForP {
		if _, ok := usedOutletIDs[o.ID]; ok {
			used++
		}
	}
	return PduSummary{
		ID:           p.ID.String(),
		Name:         p.Name,
		Side:         p.PduSide,
		Mount:        p.Mount,
		Face:         p.Face,
		TotalOutlets: len(outletsForP),
		UsedOutlets:  used,
	}
}

func usedOutletIDsFrom(conns []dbq.PowerConnection) map[uuid.UUID]struct{} {
	out := make(map[uuid.UUID]struct{}, len(conns))
	for _, c := range conns {
		out[c.OutletID] = struct{}{}
	}
	return out
}

// dedupedSides returns the sorted, deduplicated PDU-side labels seen
// in the connection list. Mirrors Python's
// `sorted({c["pdu_side"] for c in connections if c.get("pdu_side")})`.
func dedupedSides(conns []Connection) []string {
	seen := map[string]struct{}{}
	for _, c := range conns {
		if c.PduSide == nil || *c.PduSide == "" {
			continue
		}
		seen[*c.PduSide] = struct{}{}
	}
	out := make([]string, 0, len(seen))
	for s := range seen {
		out = append(out, s)
	}
	// sort lexicographically (Python's sorted())
	for i := 1; i < len(out); i++ {
		for j := i; j > 0 && out[j-1] > out[j]; j-- {
			out[j], out[j-1] = out[j-1], out[j]
		}
	}
	return out
}
